### PR TITLE
Improve ColumnarRow null handling

### PR DESCRIPTION
**Pull request by pettyjamesm**
- Adds `ColumnarRow#mayHaveNull()` and updates various usage sites to reference it when applicable
- Avoids expensive dictionaryId remapping when converting `DictionaryBlock` values into `ColumnarRow`
- Detects when `RowBlock` instances selected no null positions as part of `copyPositions` and discards null masks in those scenarios

### CODE REVIEW

**Comment by findepi** [comment]
cc @losipiuk

File: core/trino-spi/src/main/java/io/trino/spi/block/ColumnarRow.java [hunk 1]
```diff
@@ -73,8 +77,7 @@ private static ColumnarRow toColumnarRow(DictionaryBlock dictionaryBlock)
         for (int position = 0; position < dictionaryBlock.getPositionCount(); position++) {
             if (!dictionaryBlock.isNull(position)) {
                 int oldDictionaryId = dictionaryBlock.getId(position);
-                dictionaryIds[nonNullPositionCount] = newDictionaryIndex[oldDictionaryId];
-                nonNullPositionCount++;
+                dictionaryIds[nonNullPositionCount++] = newDictionaryIndex[oldDictionaryId];
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/block/ColumnarRow.java:80]: Unrelated change

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ReaderProjectionsAdapter.java [hunk 2]
```diff
@@ -150,41 +150,24 @@ private Block loadInternalBlock(List<Integer> dereferences, Block parentBlock)
...
                     newlyCreatedBlock.appendNull();
                 }
                 else {
-                    if (newlyCreatedBlock != null) {
-                        type.appendTo(loadedInternalBlock, fieldBlockPosition, newlyCreatedBlock);
-                    }
-                    fieldBlockPosition++;
+                    type.appendTo(loadedInternalBlock, fieldBlockPosition++, newlyCreatedBlock);
                 }
             }
```
Comment (martint) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/ReaderProjectionsAdapter.java:169]: Isn't it still valuable to do this? `mayHaveNull` can return `true` and the block could still contain no nulls -- after all, the only definitive answer is when it returns `false`.

File: core/trino-spi/src/main/java/io/trino/spi/block/ColumnarRow.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/block/ColumnarRow.java]: Fair enough, replaced with the original

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ReaderProjectionsAdapter.java (hunk 2 above)
Comment (pettyjamesm) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/ReaderProjectionsAdapter.java]: That's fair. I was presuming that the changes in `RowBlock` to detect no nulls present at creation time were now going to handle that scenario and would be expected to return `false` from `mayHaveNull` unless there _was_ at least one null row, but I suppose you could find a way to do a range based `RowBlock#getRegion` somewhere in between the two and end up doing unnecessary work in my revised formulation. I've added back an outer probe loop and an inner loop for once a null position is found.

**Comment by pettyjamesm** [comment]
@martint thanks for the review! I've made the requested changes, this should be ready for you to take another look when you get a chance.
